### PR TITLE
Python interface

### PR DESCRIPTION
**Pull request by hariszaf**
This is a first version of the Python interface, including the `compute_volume` and the `generate_samples` functions.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (vissarion) [inline .gitignore:1]: do you need all of those now?

File: test/volume_cb_hpolytope.cpp [hunk 2]
```diff
@@ -5,7 +5,9 @@
 
 // Licensed under GNU LGPL.3, see LICENCE file
 
-#include "doctest.h"
```
Comment (vissarion) [inline test/volume_cb_hpolytope.cpp:8]: you should include this, otherwise tests are failing

File: volestipy/setup.py [hunk 3]
```diff
@@ -1,68 +1,72 @@
...
 
 
-version = "0.2.0"
+version = "0.3.0"
 license='LGPL3',
 packages = ["volestipy"]
 description="volestipy: wrapper for the VolEsti library to sample from convex sets and compute volume."
-author = "Pedro Zuidberg Dos Martires"
-author_email="pedro.zuidbergdosmartires@cs.kuleuven.be"
+author = "Haris Zafeiropoulos"
```
Comment (vissarion) [inline volestipy/setup.py]: please do not remove the original author, just add your name

File: volestipy/setup.py [hunk 4]
```diff
@@ -1,68 +1,72 @@
...
 source_directory_list = ['volestipy', join('volestipy','src')]
 
-
 compiler_args = [
-    "-std=c++11",
-    "-O3",
-    "-lm",
-    "-ldl",
-    "-DBOOST_NO_AUTO_PTR",
+	"-std=c++11",
```
Comment (vissarion) [inline volestipy/setup.py]: is this a tab? please include only spaces tabs are evil ;) (change your settings of your IDE to replace tabs with a sequence of spaces)

File: volestipy/volestipy/include/bindings.h [hunk 5]
```diff
@@ -1,31 +1,51 @@
 #ifndef VOLESTIBINDINGS_H
 #define VOLESTIBINDINGS_H
 
-
+//// I think we could remove those...
```
Comment (vissarion) [inline volestipy/volestipy/include/bindings.h]: which one?

File: volestipy/volestipy/include/bindings.h [hunk 6]
```diff
@@ -1,31 +1,51 @@
...
       HPolytopeCPP(double *A, double *b, int n_hyperplanes, int n_variables);
 
       Hpolytope HP;
       std::pair<Point,NT> CheBall;
       ~HPolytopeCPP();
-      double compute_volume(char* method, int walk_len, double epsilon, uint seed);
-      void generate_samples(int walk_len, int n_samples, double* samples, uint seed);
+            
+      double compute_volume(char* vol_method, char* walk_method, int walk_len, double epsilon, int seed);        
+      double generate_samples(int walk_len, int number_of_points, int number_of_points_to_burn, bool boundary, bool cdhr, bool rdhr, bool gaussian, bool set_L, bool billiard, bool ball_walk, double a, double L,  double* samples);
```
Comment (vissarion) [inline volestipy/volestipy/include/bindings.h]: please wrap to avoid long lines

File: volestipy/volestipy/src/bindings.cpp [hunk 7]
```diff
@@ -3,7 +3,7 @@
 #include "bindings.h"
 
 using namespace std;
-
+//unsigned int dim;
```
Comment (vissarion) [inline volestipy/volestipy/src/bindings.cpp]: if not needed please remove

File: volestipy/volestipy/volestipy.pyx [hunk 8]
```diff
@@ -5,51 +5,67 @@
...
+   
       HPolytopeCPP() except +
-      HPolytopeCPP(double  *A, double *b, int n_varibles, int n_hyperplanes) except +
-      double compute_volume(char* method, int walk_len, double epsilon, np.npy_int32 seed)
-      double generate_samples(int walk_len, int n_samples, double* samples, np.npy_int32 seed)
+      HPolytopeCPP(double *A, double *b, int n_hyperplanes, int n_variables) except +
+
+      double compute_volume(char* vol_method, char* walk_method, int walk_len, double epsilon, int seed);     
+
+      double generate_samples(int walk_len, int number_of_points, int number_of_points_to_burn, bool boundary, bool cdhr, bool rdhr, bool gaussian, bool set_L, bool billiard, bool ball_walk, double a, double L,  double* samples);
```
Comment (vissarion) [inline volestipy/volestipy/volestipy.pyx]: long line

File: volestipy/volestipy/volestipy.pyx [hunk 9]
```diff
@@ -5,51 +5,67 @@
...
+
+
+# with respect to the "compute_volume" def
+volume_methods = ["sequence_of_balls".encode("UTF-8"), "cooling_gaussian".encode("UTF-8"), "cooling_balls".encode("UTF-8")]
+walk_methods = ["uniform_ball".encode("UTF-8"), "CDHR".encode("UTF-8"), "RDHR".encode("UTF-8"), "gaussian_ball".encode("UTF-8"), "gaussian_CDHR".encode("UTF-8"), "gaussian_RDHR".encode("UTF-8"), "uniform_ball".encode("UTF-8"), "billiard".encode("UTF-8")]
```
Comment (vissarion) [inline volestipy/volestipy/volestipy.pyx]: long line

File: volestipy/volestipy/volestipy.pyx [hunk 10]
```diff
@@ -5,51 +5,67 @@
...
-      # Do we also need to pass strdie to make it more error proof?
-      # https://stackoverflow.com/questions/34592961/passing-multidimensional-memoryviews-to-c-function
       n_hyperplanes, n_variables = A.shape[0], A.shape[1]
       self.polytope_cpp = HPolytopeCPP(&A[0,0], &b[0], n_hyperplanes, n_variables)
 
-   def compute_volume(self, int walk_len=2, double epsilon=0.05, method="sequence_of_balls", np.npy_int32 seed=get_time_seed()):
-      method = method.encode("UTF-8")
-      if method  in volume_methods:
-         return self.polytope_cpp.compute_volume(method, walk_len, epsilon, seed)
+   def compute_volume(self, walk_len=2, epsilon=0.05, vol_method="sequence_of_balls", walk_method="uniform_ball", np.npy_int32 seed=get_time_seed()):
```
Comment (vissarion) [inline volestipy/volestipy/volestipy.pyx]: long line

**Review by vissarion — COMMENTED** [review]
Thanks for this PR! I have some comments. The main issue is to include `doctest.h` again so the tests are passing.

File: volestipy/test1.py [hunk 11]
```diff
@@ -1,7 +1,12 @@
+#!/usr/bin/python3.6
+
 import numpy as np
+import volestipy
 from volestipy import HPolytope
 
 
+print("I imported everything right!")
```
Comment (TolisChal) [inline volestipy/test1.py:8]: Please remove unused comments

File: volestipy/volestipy/src/bindings.cpp [hunk 12]
```diff
@@ -24,88 +24,124 @@ HPolytopeCPP::HPolytopeCPP(double *A_np, double *b_np, int n_hyperplanes, int n_
...
+      //std::cout<<"b = "<<HP.get_vec()<<std::endl;
+      //std::cout<<"dimension = "<<HP.dimension()<<std::endl;
+      //
+      //std::cout<<"walk_len = "<<walk_len<<std::endl;
+      //std::cout<<"number_of_points = "<<number_of_points<<std::endl;
+      //std::cout<<"number_of_points_to_burn = "<<number_of_points_to_burn<<std::endl;
+      //std::cout<<"walk_len = "<<walk_len<<std::endl; 
+      //
+      //std::cout<<"a = "<<a<<std::endl;
+      //std::cout<<"L = "<<L<<std::endl;
```
Comment (TolisChal) [inline volestipy/volestipy/src/bindings.cpp]: please remove unused comments

File: volestipy/volestipy/volestipy.pyx [hunk 13]
```diff
@@ -5,51 +5,67 @@
...
+         raise Exception('"{}" is not implemented to compute volume. Available methods are: {}'.format(vol_method, volume_methods))
 
-   def generate_samples(self, int walk_len=2, int n_samples=1000, np.npy_int32 seed=get_time_seed()):
-         n_variables = self._A.shape[1]
-         cdef double[:,::1] samples = np.zeros((n_samples,  n_variables), dtype=np.float64, order="C")
-         self.polytope_cpp.generate_samples(walk_len, n_samples, &samples[0,0], seed)
-         return np.asarray(samples)
 
+# we need to build a Python function for getting a starting point depending on the polytope 
+   def generate_samples(self, walk_len=1, number_of_points=1000, number_of_points_to_burn=100, boundary=True, cdhr=True, rdhr=False, gaussian=False, set_L=False, billiard=False, ball_walk=False, a=0, L=0):
```
Comment (TolisChal) [inline volestipy/volestipy/volestipy.pyx]: set default boundary = false, and default cdhr=true

File: .gitignore (hunk 1 above)
Comment (hariszaf) [reply .gitignore:1]: I removed anything related with the ```*_old``` files.

File: test/volume_cb_hpolytope.cpp (hunk 2 above)
Comment (hariszaf) [reply test/volume_cb_hpolytope.cpp:8]: Done!

File: volestipy/setup.py (hunk 3 above)
Comment (hariszaf) [reply volestipy/setup.py]: Terribly sorry for this. I did so in the first place to distinguish the two scripts.

File: volestipy/setup.py (hunk 4 above)
Comment (hariszaf) [reply volestipy/setup.py]: Done! :)

File: volestipy/test1.py (hunk 11 above)
Comment (hariszaf) [reply volestipy/test1.py:8]: Done!

File: volestipy/volestipy/include/bindings.h (hunk 5 above)
Comment (hariszaf) [reply volestipy/volestipy/include/bindings.h]: I think we could remove the following:
``` 
#include "Eigen/Eigen"
#include "cartesian_kernel.h" 
```
without them, volestipy still compliles and runs ```test1.py``` and ```test2.py```

Fo now, I have just comment them out. Any suggestions? :)

File: volestipy/volestipy/include/bindings.h (hunk 6 above)
Comment (hariszaf) [reply volestipy/volestipy/include/bindings.h]: Done!

File: volestipy/volestipy/src/bindings.cpp (hunk 7 above)
Comment (hariszaf) [reply volestipy/volestipy/src/bindings.cpp]: Done! It has been removed.

File: volestipy/volestipy/src/bindings.cpp (hunk 12 above)
Comment (hariszaf) [reply volestipy/volestipy/src/bindings.cpp]: Ok!

File: volestipy/volestipy/volestipy.pyx (hunk 8 above)
Comment (hariszaf) [reply volestipy/volestipy/volestipy.pyx]: Line break added

File: volestipy/volestipy/volestipy.pyx (hunk 9 above)
Comment (hariszaf) [reply volestipy/volestipy/volestipy.pyx]: Line break added

File: volestipy/volestipy/volestipy.pyx (hunk 10 above)
Comment (hariszaf) [reply volestipy/volestipy/volestipy.pyx]: Line break added

File: volestipy/volestipy/volestipy.pyx (hunk 13 above)
Comment (hariszaf) [reply volestipy/volestipy/volestipy.pyx]: boundary changed 
cdhr was true

File: volestipy/volestipy/include/bindings.h (hunk 5 above)
Comment (vissarion) [reply volestipy/volestipy/include/bindings.h]: probably not needed, you can remove them

File: .gitignore (hunk 1 above)
Comment (vissarion) [reply .gitignore:1]: do you need those too
```
test/Testing/Temporary/LastTest.log
test/Testing/Temporary/CTestCostData.txt
```
are the created by cython or something?

File: test/volume_cb_hpolytope.cpp (hunk 2 above)
Comment (vissarion) [reply test/volume_cb_hpolytope.cpp:8]: tests are passing now, great!

File: volestipy/setup.py [hunk 14]
```diff
@@ -1,75 +1,83 @@
...
 
-version = "0.2.0"
+version = "0.3.0"
 license='LGPL3',
 packages = ["volestipy"]
 description="volestipy: wrapper for the VolEsti library to sample from convex sets and compute volume."
-author = "Pedro Zuidberg Dos Martires"
-author_email="pedro.zuidbergdosmartires@cs.kuleuven.be"
+author = "Pedro Zuidberg Dos Martires, Haris Zafeiropoulos"
+author_email="pedro.zuidbergdosmartires@cs.kuleuven.be, haris-zaf@hcmr.gr"
```
Comment (vissarion) [inline volestipy/setup.py:15]: I think it is ok for now, but also read this https://bugs.python.org/msg93284

File: volestipy/setup.py [hunk 15]
```diff
@@ -1,75 +1,83 @@
...
 
 
-version = "0.2.0"
+version = "0.3.0"
 license='LGPL3',
 packages = ["volestipy"]
 description="volestipy: wrapper for the VolEsti library to sample from convex sets and compute volume."
-author = "Pedro Zuidberg Dos Martires"
-author_email="pedro.zuidbergdosmartires@cs.kuleuven.be"
+author = "Pedro Zuidberg Dos Martires, Haris Zafeiropoulos"
```
Comment (vissarion) [inline volestipy/setup.py:14]: you can add your name in credits if you like https://github.com/GeomScale/volume_approximation/blob/develop/doc/credits.md

Comment (hariszaf) [reply volestipy/setup.py:14]: Ok. I did that. :)

File: volestipy/setup.py (hunk 14 above)
Comment (hariszaf) [reply volestipy/setup.py:15]: Ok! Thanks for sharing this!

File: .gitignore (hunk 1 above)
Comment (hariszaf) [reply .gitignore:1]: These were there in the initial ```.gitignore``` file. 
Would you like to remove them?

Comment (vissarion) [reply .gitignore:1]: Yes, if you check that those files are not generated locally on your side, which is my guess.

**Pull request merged**